File: backend/app/repositories/run_repository.py
"""
Run Repository
Handles all database operations for workflow runs
"""
from typing import List, Optional, Dict, Any
from datetime import datetime, UTC
import uuid

from app.models import Run, RunCreate, RunResult


class RunRepository:
    """Repository for Run CRUD operations"""
    
    @staticmethod
    async def create(run_data: RunCreate) -> Run:
        """Create a new workflow run"""
        run = Run(
            runId=str(uuid.uuid4()),
            workflowId=run_data.workflowId,
            status="pending",
            trigger="manual",
            variables=run_data.variables or {},
            callbackUrl=run_data.callbackUrl,
            results=[],
            createdAt=datetime.now(UTC),
            startedAt=None,
            completedAt=None,
            duration=None,
            error=None
        )
        
        await run.insert()
        return run
    
    @staticmethod
    async def get_by_id(run_id: str) -> Optional[Run]:
        """Get run by runId - SQL injection safe"""
        return await Run.find_one(Run.runId == run_id)
    
    @staticmethod
    async def list_by_workflow(
        workflow_id: str,
        skip: int = 0,
        limit: int = 20
    ) -> tuple[List[Run], int]:
        """Get runs for a specific workflow"""
        query = Run.find(Run.workflowId == workflow_id)
        
        total = await query.count()
        runs = await query.sort(-Run.createdAt).skip(skip).limit(limit).to_list()
        
        return runs, total
    
    @staticmethod
    async def list_all(skip: int = 0, limit: int = 20) -> tuple[List[Run], int]:
        """List all runs with pagination"""
        total = await Run.count()
        runs = await Run.find_all().sort(-Run.createdAt).skip(skip).limit(limit).to_list()
        
        return runs, total
    
    @staticmethod
    async def update_status(
        run_id: str,
        status: str,
        error: Optional[str] = None
    ) -> Optional[Run]:
        """Update run status"""
        run = await RunRepository.get_by_id(run_id)
        if not run:
            return None
        
        run.status = status
        if error:
            run.error = error
        
        if status == "running" and not run.startedAt:
            run.startedAt = datetime.now(UTC)
        
        if status in ["completed", "failed", "cancelled"]:
            run.completedAt = datetime.now(UTC)
            if run.startedAt:
                run.duration = int((run.completedAt - run.startedAt).total_seconds() * 1000)
        
        await run.save()
        return run
    
    @staticmethod
    async def add_result(run_id: str, result: RunResult) -> Optional[Run]:
        """Add a node execution result to the run"""
        run = await RunRepository.get_by_id(run_id)
        if not run:
            return None
        
        run.results.append(result)
        await run.save()
        
        return run
    
    @staticmethod
    async def update_results(run_id: str, results: List[RunResult]) -> Optional[Run]:
        """Update all results for a run"""
        run = await RunRepository.get_by_id(run_id)
        if not run:
            return None
        
        run.results = results
        await run.save()
        
        return run
    
    @staticmethod
    async def delete(run_id: str) -> bool:
        """Delete a run"""
        run = await RunRepository.get_by_id(run_id)
        if not run:
            return False
        
        await run.delete()
        return True
    
    @staticmethod
    async def delete_by_workflow(workflow_id: str) -> int:
        """Delete all runs for a workflow - returns count deleted"""
        result = await Run.find(Run.workflowId == workflow_id).delete()
        return result.deleted_count if result else 0
    
    @staticmethod
    async def get_recent_runs(limit: int = 10) -> List[Run]:
        """Get most recent runs across all workflows"""
        return await Run.find_all().sort(-Run.createdAt).limit(limit).to_list()

    @staticmethod
    async def get_latest_failed_run(workflow_id: str) -> Optional[Run]:
        """Get latest failed run for a workflow"""
        return await Run.find(
            Run.workflowId == workflow_id,
            Run.status == "failed",
        ).sort(-Run.createdAt).first_or_none()

    @staticmethod
    async def get_latest_run(workflow_id: str) -> Optional[Run]:
        """Get latest run for a workflow regardless of status"""
        return await Run.find(
            Run.workflowId == workflow_id,
        ).sort(-Run.createdAt).first_or_none()
    
    @staticmethod
    async def count_by_status(workflow_id: Optional[str] = None) -> Dict[str, int]:
        """Get count of runs by status"""
        query = Run.find(Run.workflowId == workflow_id) if workflow_id else Run.find_all()
        
        all_runs = await query.to_list()
        
        counts = {
            "pending": 0,
            "running": 0,
            "completed": 0,
            "failed": 0,
            "cancelled": 0
        }
        
        for run in all_runs:
            counts[run.status] = counts.get(run.status, 0) + 1
        
        return counts
    
    @staticmethod
    async def update_fields(run_id: str, **fields) -> Optional[Run]:
        """Update specific fields of a run"""
        run = await RunRepository.get_by_id(run_id)
        if not run:
            return None
        
        for field, value in fields.items():
            if hasattr(run, field):
                setattr(run, field, value)
        
        await run.save()
        return run

File: desktop/core/repositories/RunRepository.ts
import type { KVStore, SqliteRow } from "../db"
import { SIDE_TABLE_THRESHOLD_BYTES } from "../db"
import type { Run } from "../../../shared/types/Run"
import type { RunResult } from "../../../shared/types/RunResult"
import type { JsonValue } from "../../../shared/types/JsonValue"
import { generateId } from "../id"
import { mustExist, parseJson, toJson } from "./helpers"

export type RunCreate = Pick<Run, "workspaceId" | "workflowId"> &
  Partial<Pick<Run, "status" | "trigger" | "variables" | "selectedEnvironmentId" | "nodeStatuses">>

export type RunUpdate = Partial<
  Pick<
    Run,
    | "status"
    | "trigger"
    | "variables"
    | "nodeStatuses"
    | "results"
    | "selectedEnvironmentId"
    | "startedAt"
    | "completedAt"
    | "duration"
    | "error"
    | "failedNodes"
    | "failureMessage"
    | "resumeFromRunId"
    | "resumeFromNodeIds"
    | "resumeMode"
  >
>

/** Where a persisted node-response body ended up. */
export type BodyStorage = "inline" | "side"

const TERMINAL_STATUSES: ReadonlySet<Run["status"]> = new Set(["completed", "failed", "cancelled", "interrupted"])

const COLUMNS =
  "id, workspace_id, workflow_id, status, node_statuses_json, extracted_variables_json, response_metadata_json, startedAt, completedAt, rev, createdAt, updatedAt"

interface RunRow extends SqliteRow {
  readonly id: string
  readonly workspace_id: string
  readonly workflow_id: string
  readonly status: string
  readonly node_statuses_json: string
  readonly extracted_variables_json: string
  readonly response_metadata_json: string
  readonly startedAt: string | null
  readonly completedAt: string | null
  readonly rev: number
  readonly createdAt: string
  readonly updatedAt: string
}

/** Everything on a Run that has no dedicated column rides in this blob. */
interface RunMetadata {
  readonly selectedEnvironmentId: string | null
  readonly trigger: Run["trigger"]
  readonly results: readonly RunResult[]
  readonly duration: number | null
  readonly error: string | null
  readonly failedNodes: readonly string[] | null
  readonly failureMessage: string | null
  readonly resumeFromRunId: string | null
  readonly resumeFromNodeIds: readonly string[] | null
  readonly resumeMode: "single" | "all-failed" | null
}

export class RunRepository {
  public constructor(private readonly store: KVStore) {}

  public create(input: RunCreate): Run {
    const id = generateId()
    const metadata: RunMetadata = {
      selectedEnvironmentId: input.selectedEnvironmentId ?? null,
      trigger: input.trigger ?? "manual",
      results: [],
      duration: null,
      error: null,
      failedNodes: null,
      failureMessage: null,
      resumeFromRunId: null,
      resumeFromNodeIds: null,
      resumeMode: null,
    }
    this.store.set(
      "INSERT INTO runs (id, workspace_id, workflow_id, scopeId, status, node_statuses_json, extracted_variables_json, response_metadata_json) VALUES (?, ?, ?, ?, ?, ?, ?, ?)",
      [
        id,
        input.workspaceId,
        input.workflowId,
        input.workspaceId,
        input.status ?? "pending",
        toJson(input.nodeStatuses ?? {}),
        toJson(input.variables ?? {}),
        toJson(metadata),
      ],
    )
    return mustExist(this.getById(id), `run ${id} missing after insert`)
  }

  public getById(runId: string): Run | undefined {
    const row = this.store.get<RunRow>(`SELECT ${COLUMNS} FROM runs WHERE id = ?`, [runId])
    return row === undefined ? undefined : rowToRun(row)
  }

  public listByWorkflow(workflowId: string): { items: readonly Run[]; total: number } {
    const items = this.store
      .query<RunRow>(`SELECT ${COLUMNS} FROM runs WHERE workflow_id = ? ORDER BY createdAt DESC, id DESC`, [workflowId])
      .map(rowToRun)
    return { items, total: items.length }
  }

  public listByWorkspace(workspaceId: string): { items: readonly Run[]; total: number } {
    const items = this.store
      .query<RunRow>(`SELECT ${COLUMNS} FROM runs WHERE workspace_id = ? ORDER BY createdAt DESC, id DESC`, [workspaceId])
      .map(rowToRun)
    return { items, total: items.length }
  }

  public getLatestRun(workflowId: string): Run | undefined {
    const row = this.store.get<RunRow>(
      `SELECT ${COLUMNS} FROM runs WHERE workflow_id = ? ORDER BY createdAt DESC, id DESC LIMIT 1`,
      [workflowId],
    )
    return row === undefined ? undefined : rowToRun(row)
  }

  public getLatestFailedRun(workflowId: string): Run | undefined {
    const row = this.store.get<RunRow>(
      `SELECT ${COLUMNS} FROM runs WHERE workflow_id = ? AND status = 'failed' ORDER BY createdAt DESC, id DESC LIMIT 1`,
      [workflowId],
    )
    return row === undefined ? undefined : rowToRun(row)
  }

  public update(runId: string, patch: RunUpdate): Run | undefined {
    const existing = this.getById(runId)
    if (existing === undefined) {
      return undefined
    }
    this.writeRun({ ...existing, ...patch })
    return this.getById(runId)
  }

  /**
   * Port of `RunRepository.update_status`: stamps `startedAt` on first
   * transition to running and `completedAt`/`duration` on any terminal state.
   */
  public updateStatus(runId: string, status: Run["status"], error?: string): Run | undefined {
    const existing = this.getById(runId)
    if (existing === undefined) {
      return undefined
    }
    const now = new Date().toISOString()
    const startedAt = status === "running" && existing.startedAt == null ? now : existing.startedAt ?? null
    const terminal = TERMINAL_STATUSES.has(status)
    const completedAt = terminal ? now : existing.completedAt ?? null
    const duration =
      terminal && startedAt != null ? Date.parse(completedAt ?? now) - Date.parse(startedAt) : existing.duration ?? null
    this.writeRun({
      ...existing,
      status,
      error: error ?? existing.error ?? null,
      startedAt,
      completedAt,
      duration,
    })
    return this.getById(runId)
  }

  public updateResults(runId: string, results: readonly RunResult[]): Run | undefined {
    return this.update(runId, { results: [...results] })
  }

  public delete(runId: string): boolean {
    return this.store.delete("DELETE FROM runs WHERE id = ?", [runId]).changes > 0
  }

  public deleteByWorkflow(workflowId: string): number {
    return this.store.delete("DELETE FROM runs WHERE workflow_id = ?", [workflowId]).changes
  }

  /**
   * Persist a node's response body. Bodies at/above the side-table threshold
   * spill to `run_responses` (decision #7); smaller ones stay inline in the
   * run's results and are NOT written here.
   */
  public putNodeBody(runId: string, nodeId: string, body: Buffer): BodyStorage {
    if (body.length < SIDE_TABLE_THRESHOLD_BYTES) {
      return "inline"
    }
    this.store.set(
      "INSERT INTO run_responses (run_id, node_id, body, size) VALUES (?, ?, ?, ?) " +
        "ON CONFLICT(run_id, node_id) DO UPDATE SET body = excluded.body, size = excluded.size",
      [runId, nodeId, body, body.length],
    )
    return "side"
  }

  public getNodeBody(runId: string, nodeId: string): Buffer | undefined {
    const row = this.store.get<{ body: Buffer } & SqliteRow>(
      "SELECT body FROM run_responses WHERE run_id = ? AND node_id = ?",
      [runId, nodeId],
    )
    return row?.body
  }

  private writeRun(run: Run): void {
    const metadata: RunMetadata = {
      selectedEnvironmentId: run.selectedEnvironmentId ?? null,
      trigger: run.trigger,
      results: run.results,
      duration: run.duration ?? null,
      error: run.error ?? null,
      failedNodes: run.failedNodes ?? null,
      failureMessage: run.failureMessage ?? null,
      resumeFromRunId: run.resumeFromRunId ?? null,
      resumeFromNodeIds: run.resumeFromNodeIds ?? null,
      resumeMode: run.resumeMode ?? null,
    }
    this.store.set(
      "UPDATE runs SET status = ?, node_statuses_json = ?, extracted_variables_json = ?, response_metadata_json = ?, startedAt = ?, completedAt = ? WHERE id = ?",
      [
        run.status,
        toJson(run.nodeStatuses),
        toJson(run.variables),
        toJson(metadata),
        run.startedAt ?? null,
        run.completedAt ?? null,
        run.runId,
      ],
    )
  }
}

function rowToRun(row: RunRow): Run {
  const metadata = parseJson<RunMetadata>(row.response_metadata_json)
  return {
    runId: row.id,
    workspaceId: row.workspace_id,
    workflowId: row.workflow_id,
    selectedEnvironmentId: metadata.selectedEnvironmentId,
    status: row.status as Run["status"],
    trigger: metadata.trigger,
    variables: parseJson<Record<string, JsonValue>>(row.extracted_variables_json),
    results: [...metadata.results],
    startedAt: row.startedAt,
    completedAt: row.completedAt,
    duration: metadata.duration,
    error: metadata.error,
    failedNodes: metadata.failedNodes === null ? null : [...metadata.failedNodes],
    failureMessage: metadata.failureMessage,
    nodeStatuses: parseJson<Record<string, JsonValue>>(row.node_statuses_json),
    resumeFromRunId: metadata.resumeFromRunId,
    resumeFromNodeIds: metadata.resumeFromNodeIds === null ? null : [...metadata.resumeFromNodeIds],
    resumeMode: metadata.resumeMode,
    rev: row.rev,
    createdAt: row.createdAt,
    updatedAt: row.updatedAt,
  }
}
